ecl/hql/hqlopt.cpp

```diff
@@ -3988,6 +3983,34 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
                     }
                     break;
                 }
+            case no_if:
+                {
+                    if (transformed->queryChild(1)->isConstant())
+                        return swapIntoIf(transformed);
+                    break;
+                }
+            }
+            break;
+        }
+    case no_select:
+        {
+            node_operator childOp = child->getOperator();
+            switch(childOp)
+            {
+                case no_if:
+                {
+                      if (!isShared(child) && child->isDatarow())
+                      {
+                          IHqlExpression * cond = child->queryChild(0);
+                          IHqlExpression * field = transformed->queryChild(1);
+                          OwnedHqlExpr newLeft = createSelectExpr(LINK(child->queryChild(1)), LINK(field));
```

Comment:
I suspect this should be calling createNewSelectExpr() - otherwise if child->queryChild(1) is a complex expression it will not be evaluated properly.